### PR TITLE
sl_3_z_ala_fp_groups

### PR DESCRIPTION
**Pull request by piotrmizerka**

### CODE REVIEW

**Comment by piotrmizerka** [comment]
For Friday: test gersten relations

**Comment by piotrmizerka** [comment]
@kalmarek I noticed there is a bug in this function:
```julia
function certify_sos_decomposition(
    X,
    order_unit,
    λ::Number,
    Q::AbstractMatrix,
    support,
    RG::StarAlgebra,
)
    λ_interval = @interval(λ)
    eoi = X - λ_interval * order_unit

    residual = eoi - sos_from_matrix(Q, support, RG)
    l1_norm = sum(x -> norm(x, 1), residual)

    @info "l₁ norm of the error in interval arithmetic:" l1_norm radius(l1_norm)

    result = λ_interval - l1_norm

    return result
end
```

This bug resides in the line `eoi = X - λ_interval * order_unit`. I think it worked previously, because in the former implementation, our group ring was defined by matrices and without using of PropertyT_new way for SL(n,Z). Now there is a problem with this. I think this boils down to the problem with multiplying group ring elements by intervals for group rings defined by FPGroup types groups. I checked this for G = FPGroup = C_3 and it didn't work either.

**Comment by kalmarek** [comment]
What is the problem? what is the error message?

**Comment by piotrmizerka** [comment]
`certify_sos_decomposition: Error During Test at /home/piotr/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/certification_tests.jl:59
  Got exception outside of a @test
  MethodError: Cannot `convert` an object of type 
    AlgebraElement{StarAlgebra{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},StarAlgebras.CachedMTable{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},StarAlgebras.Basis{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},Array{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},1}},Array{Int64{},2},true},StarAlgebras.Basis{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},Array{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},1}}},Interval{Float64},SparseArrays.SparseVector{Interval{Float64},Int64{}}} to an object of type 
    AlgebraElement{StarAlgebra{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},StarAlgebras.CachedMTable{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},StarAlgebras.Basis{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},Array{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},1}},Array{Int64{},2},true},StarAlgebras.Basis{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},Array{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},1}}},Float64,SparseArrays.SparseVector{Float64,Int64{}}}
  Closest candidates are:
    convert(::Type{T}, ::T) where T at essentials.jl:205
  Stacktrace:
    [1] setindex!
      @ ./array.jl:841 [inlined]
    [2] setindex!
      @ ./multidimensional.jl:639 [inlined]
    [3] macro expansion
      @ ./broadcast.jl:984 [inlined]
    [4] macro expansion
      @ ./simdloop.jl:77 [inlined]
    [5] copyto!
      @ ./broadcast.jl:983 [inlined]
    [6] copyto!
      @ ./broadcast.jl:936 [inlined]
    [7] copy
      @ ./broadcast.jl:908 [inlined]
    [8] materialize
      @ ./broadcast.jl:883 [inlined]
    [9] broadcast_preserving_zero_d
      @ ./broadcast.jl:872 [inlined]
   [10] *(A::Interval{Float64}, B::Matrix{AlgebraElement{StarAlgebra{FPGroup{Symbol, …}, FPGroupElement{FPGroup{Symbol, …}, …}, …}, Int64, SparseArrays.SparseVector{Int64, Int64}}})
      @ Base ./arraymath.jl:52
   [11] certify_sos_decomposition(X::Matrix{AlgebraElement{StarAlgebra{FPGroup{Symbol, …}, FPGroupElement{FPGroup{Symbol, …}, …}, …}, Int64, SparseArrays.SparseVector{Int64, Int64}}}, order_unit::Matrix{AlgebraElement{StarAlgebra{FPGroup{Symbol, …}, FPGroupElement{FPGroup{Symbol, …}, …}, …}, Int64, SparseArrays.SparseVector{Int64, Int64}}}, λ::Int64, Q::Matrix{Int64}, support::Vector{FPGroupElement{FPGroup{Symbol, …}, …}}, RG::StarAlgebra{FPGroup{Symbol, …}, FPGroupElement{FPGroup{Symbol, …}, …}, …})
      @ LowCohomologySOS ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/src/certification.jl:31
   [12] (::var"#58#61"{Matrix{AlgebraElement{StarAlgebra{FPGroup{Symbol, …}, FPGroupElement{FPGroup{Symbol, …}, …}, …}, Int64, SparseArrays.SparseVector{Int64, Int64}}}, Matrix{AlgebraElement{StarAlgebra{FPGroup{Symbol, …}, FPGroupElement{FPGroup{Symbol, …}, …}, …}, Int64, SparseArrays.SparseVector{Int64, Int64}}}})()
      @ Main ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/certification_tests.jl:65
   [13] with_logstate(f::Function, logstate::Any)
      @ Base.CoreLogging ./logging.jl:491
   [14] with_logger(f::Function, logger::Base.CoreLogging.NullLogger)
      @ Base.CoreLogging ./logging.jl:603
   [15] macro expansion
      @ ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/certification_tests.jl:64 [inlined]
   [16] macro expansion
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [17] top-level scope
      @ ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/certification_tests.jl:60
   [18] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [19] macro expansion
      @ ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/runtests.jl:51 [inlined]
   [20] macro expansion
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [21] top-level scope
      @ ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/runtests.jl:48
   [22] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [23] top-level scope
      @ none:6
   [24] eval
      @ ./boot.jl:360 [inlined]
   [25] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:261
   [26] _start()
      @ Base ./client.jl:485
spectral_gaps_certification: Error During Test at /home/piotr/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/certification_tests.jl:114
  Got exception outside of a @test
  MethodError: Cannot `convert` an object of type 
    AlgebraElement{StarAlgebra{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},StarAlgebras.CachedMTable{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},StarAlgebras.Basis{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},Array{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},1}},Array{Int64{},2},true},StarAlgebras.Basis{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},Array{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},1}}},Interval{Float64},SparseArrays.SparseVector{Interval{Float64},Int64{}}} to an object of type 
    AlgebraElement{StarAlgebra{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},StarAlgebras.CachedMTable{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},StarAlgebras.Basis{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},Array{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},1}},Array{Int64{},2},true},StarAlgebras.Basis{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},Int64{},Array{FPGroupElement{FPGroup{Symbol{},KnuthBendix.RewritingSystem{KnuthBendix.Word{UInt8{}},KnuthBendix.LenLex{Symbol{}}},FPGroupElement{FreeGroup{Symbol{},KnuthBendix.LenLex{Symbol{}}},KnuthBendix.Word{UInt8{}}}},KnuthBendix.Word{UInt8{}}},1}}},Float64,SparseArrays.SparseVector{Float64,Int64{}}}
  Closest candidates are:
    convert(::Type{T}, ::T) where T at essentials.jl:205
  Stacktrace:
    [1] setindex!
      @ ./array.jl:841 [inlined]
    [2] setindex!
      @ ./multidimensional.jl:639 [inlined]
    [3] macro expansion
      @ ./broadcast.jl:984 [inlined]
    [4] macro expansion
      @ ./simdloop.jl:77 [inlined]
    [5] copyto!
      @ ./broadcast.jl:983 [inlined]
    [6] copyto!
      @ ./broadcast.jl:936 [inlined]
    [7] copy
      @ ./broadcast.jl:908 [inlined]
    [8] materialize
      @ ./broadcast.jl:883 [inlined]
    [9] broadcast_preserving_zero_d
      @ ./broadcast.jl:872 [inlined]
   [10] *(A::Interval{Float64}, B::Matrix{AlgebraElement{StarAlgebra{FPGroup{Symbol, …}, FPGroupElement{FPGroup{Symbol, …}, …}, …}, Int64, SparseArrays.SparseVector{Int64, Int64}}})
      @ Base ./arraymath.jl:52
   [11] certify_sos_decomposition(X::Matrix{AlgebraElement{StarAlgebra{FPGroup{Symbol, …}, FPGroupElement{FPGroup{Symbol, …}, …}, …}, Int64, SparseArrays.SparseVector{Int64, Int64}}}, order_unit::Matrix{AlgebraElement{StarAlgebra{FPGroup{Symbol, …}, FPGroupElement{FPGroup{Symbol, …}, …}, …}, Int64, SparseArrays.SparseVector{Int64, Int64}}}, λ::Float64, Q::LinearAlgebra.Symmetric{Float64, Matrix{Float64}}, support::Vector{FPGroupElement{FPGroup{Symbol, …}, …}}, RG::StarAlgebra{FPGroup{Symbol, …}, FPGroupElement{FPGroup{Symbol, …}, …}, …})
      @ LowCohomologySOS ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/src/certification.jl:31
   [12] spectral_gaps_certification(h::Function, relations::Vector{FPGroupElement{FreeGroup{Symbol, KnuthBendix.LenLex{Symbol}}, …}}, half_basis::Vector{FPGroupElement{FPGroup{Symbol, …}, …}}; optimizer::MathOptInterface.OptimizerWithAttributes)
      @ LowCohomologySOS ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/src/certification.jl:60
   [13] (::var"#64#66"{Vector{FPGroupElement{FreeGroup{Symbol, KnuthBendix.LenLex{Symbol}}, …}}, var"#quotient_hom_1#65"{FPGroupElement{FreeGroup{Symbol, KnuthBendix.LenLex{Symbol}}, …}, FreeGroup{Symbol, KnuthBendix.LenLex{Symbol}}}})()
      @ Main ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/certification_tests.jl:132
   [14] with_logstate(f::Function, logstate::Any)
      @ Base.CoreLogging ./logging.jl:491
   [15] with_logger(f::Function, logger::Base.CoreLogging.NullLogger)
      @ Base.CoreLogging ./logging.jl:603
   [16] macro expansion
      @ ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/certification_tests.jl:131 [inlined]
   [17] macro expansion
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [18] top-level scope
      @ ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/certification_tests.jl:115
   [19] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [20] macro expansion
      @ ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/runtests.jl:51 [inlined]
   [21] macro expansion
      @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [22] top-level scope
      @ ~/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/runtests.jl:48
   [23] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [24] top-level scope
      @ none:6
   [25] eval
      @ ./boot.jl:360 [inlined]
   [26] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:261
   [27] _start()
      @ Base ./client.jl:485
Test Summary:                 | Pass  Error  Total
LowCohomologySOS              |  152      2    154
  cyclic_group                |    2             2
  group_ring                  |   20            20
  d₀                          |    2             2
  embed                       |   22            22
  fox_derivative              |   12            12
  jacobian_matrix             |    6             6
  suitable_group_ring         |   10            10
  constraints                 |    3             3
  entry_constraint            |    4             4
  sos_problem_matrix          |    4             4
  sos_problem_solution        |    4             4
  spectral_gaps_approximated  |    4             4
  sos_from_matrix             |    3             3
  certify_sos_decomposition   |           1      1
  spectral_gaps_certification |           1      1
  SL(3, ℤ)                    |   56            56
ERROR: LoadError: Some tests did not pass: 152 passed, 0 failed, 2 errored, 0 broken.
in expression starting at /home/piotr/Desktop/postdoc-warsaw/code/LowCohomologySOS/test/runtests.jl:47
ERROR: Package LowCohomologySOS errored during testing`

**Comment by piotrmizerka** [comment]
@kalmarek Friday at 14 is ok for me, we can meet then ;)

File: Project.toml [hunk 1]
```diff
@@ -13,6 +13,7 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 Logging = "56ddb016-857b-54e1-b83d-db4d58db5568"
 MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
 PropertyT_new = "8df21a36-ccc5-57bd-9352-45f2918782cf"
+Revise = "295af30f-e4ad-537b-8983-00126c2a3abe"
```
Comment (kalmarek) [inline Project.toml]: Revise should sit in your global julia env, not here ;)

**Pull request merged**